docs(api): tutorial.rst revisions for Flex

jwwojak opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

Suggested revisions to the [Tutorial docs](https://docs.opentrons.com/v2/tutorial.html) (`tutorial.rst` file) to include Flex code and examples. Changes include:

- Minor text revisions.
- New section "Requirements" covering the `requirements` code block.
- Using the Sphinx `..tabs::` directive to include and display OT-2 and Flex code samples and text in a compact manner.

# Test Plan

Verify that the API load name for the Flex pipette is accurate.

Requesting reviews.

# Changelog

**Hardware and Labware section**
Original uses 3 bullets and with nested bullets under a labware bullet. I think we can get rid of the nested bullets.

**Create a Protocol File**
Suggesting revision to the 1st paragraph after the `opentrons import` code sample. Trying to simplify it.

**Metadata**
Added a "Note" at the end about using the `requirements` code block.

**Requirements**
New section describes the `requirements` code block with Flex and OT-2 examples. 

**Labware**
Using tabs to display similar Flex and OT-2 information. Needs Flex deck image.

**Pipettes**
Add Flex code sample/example.

**Commands**
Add Flex code sample/example. Not updating the animated gif as it's valid for both robots.

**8-Channel Pipette**
Add Flex code sample/example.

# Review requests

- Check Flex code for accuracy.
- Give the Requirements section some attention. The info there is my best understanding of what this code block is and when it is needed. It's written as a hard requirement for the conditions described, but may not be an absolute, hard requirement (putting the API version in `requirements` instead of `metadata`).

# Risk assessment

Low. This is a change to a single docs section/page. It doesn't touch other production code.

jwwojak committed:
Revisions to tutorial.rst file

This starts the process of revising the API docs to include the Flex and OT-2. Specifically, we're attempting  to update the tutorial.

jwwojak committed:
Update tutorial.rst

jwwojak committed:
Update tutorial.rst

Adding some tabs. Committing to save and also as a test.

jwwojak committed:
Update tutorial.rst

Saving text, code, and tab formatting.

jwwojak committed:
Merge branch 'edge' into tutorial-page-revisions

jwwojak committed:
Update tutorial.rst

Continuing work. Saving before lunch.

jwwojak committed:
Update tutorial.rst

Final check and text changes.

Also, can't find Flex load_name for pipette for the Pipette section. Checking on that and substituting Xs for a name. Example:

XXXX = protocol.load_instrument('LOAD_NAME_PLACEHOLDER', 'left', tip_racks[tips])

And needs Flex deck map image.

jwwojak committed:
Update tutorial.rst

Adding a comment about how OT-2 code in animation also applies to Flex.

jwwojak committed:
Update tutorial.rst

jwwojak committed:
Update tutorial.rst

Adding Flex 1-channel, 1000 ul pipette load name.

jwwojak committed:
Update tutorial.rst

modifying comment.

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -21,15 +20,22 @@ The lab task that you’ll automate in this tutorial is `serial dilution`: takin
+.. Placeholders for Flex content that needs links to online manual.
 
-- `Set up your OT-2 <https://support.opentrons.com/s/ot2-get-started>`_, including `attaching pipettes <https://support.opentrons.com/s/article/Get-started-Attach-pipettes>`_ and `initial calibration <https://support.opentrons.com/s/article/Get-started-Calibrate-the-deck>`_. You can use either a single-channel or 8-channel pipette for this tutorial. Most code examples will use a `P300 Single-Chan ...
+- **Flex users** should review Chapter 2: Installation and Relocation (**LINK TBD**) in the instruction manual. Specifically, see the pipette information in the Instrument Installation and Calibration section (**LINK TBD**). You can use either a 1-channel or 8-channel pipette for this tutorial. Most Flex code examples will use a `1-channel, 1000 uL pipette <https://opentrons.com/products/flex-pip ...
```
Let's just link `instruction manual` for now. And to mirror what we have for OT-2 (and to be as specific as possible), we can link to the shop page for the 1-channel. I've also updated it to use the official short name. 
```suggestion
- **Flex users** should review Chapter 2: Installation and Relocation in the `instruction manual <https://insights.opentrons.com/hubfs/Products/Flex/Opentrons%20Flex%20Manual.pdf>`_. Specifically, see the pipette information in the Instrument Installation and Calibration section. You can use either a 1-channel or 8-channel pipette for this tutorial. Most Flex code examples will use a `Flex 1-Channel 1000 μL pipette <https://shop.opentrons.com/opentrons-flex-1-channel-pipette/>`_.
```

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -21,15 +20,22 @@ The lab task that you’ll automate in this tutorial is `serial dilution`: takin
-- Prepare the labware you need for serial dilution. The code examples will use the following labware definitions, but as long as you have labware of each type you can modify the code to run with your labware.
+- **OT-2 users** should review the robot setup and pipette information on the `Get Started page <https://support.opentrons.com/s/ot2-get-started>`_. Specifically, see `attaching pipettes <https://support.opentrons.com/s/article/Get-started-Attach-pipettes>`_ and `initial calibration <https://support.opentrons.com/s/article/Get-started-Calibrate-the-deck>`_. You can use either a single-channel or  ...
+
+The Flex and OT-2 use the same labware for serial dilution. The tutorial code will use the following labware definitions, but as long as you have labware of each type you can modify the code to run with your labware.
```
the tip racks are different for Flex/OT-2. we should either factor that out to the robot-specific sections above or load all that info into the one bullet below.

jwwojak replied:
Ohhhhh, right! The 200 ul Flex tips are in the shop site but not on the labware page. URLs for the Flex and OT-2 will be different: Flex goes to "shop.opentrons.com" and OT-2 will go to "labware.opentrons.com". 

We'll use the correct API load name for each: 

- opentrons_flex_96_tiprack_200ul
- opentrons_96_tiprack_300ul

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -46,13 +52,13 @@ Let’s start from scratch to create your serial dilution protocol. Open up a ne
+.. Suggest simplifying this paragraph
 
-Everything else in the protocol file will be required. Next, you’ll specify the version of the API you’re using. Then comes the core of the protocol: defining a single ``run()`` function that provides the locations of your labware, states which kind of pipettes you’ll use, and finally issues the commands that the robot will perform.
+Throughout this documentation, you’ll see protocols that begin with the ``opentrons import`` statement shown above. It identifies your code as an Opentrons protocol. This statement is not required, but including it is a good practice and allows most code editors to provide helpful autocomplete suggestions. 
```
`opentrons import` is not idiomatic for python users. either
```the ``import`` statement```
or
```the ``from opentrons import`` statement```

jwwojak replied:
Will replace with "the `import` statement" as in the original.

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -46,13 +52,13 @@ Let’s start from scratch to create your serial dilution protocol. Open up a ne
     from opentrons import protocol_api
 
-Throughout this documentation you’ll see protocols that begin this way. It’s not strictly necessary to include this ``import`` statement — you could leave it out and your protocol would still run fine — but it indicates clearly that this is an Opentrons protocol and, depending on your editor, it may provide you with autocomplete suggestions or other niceties.
+.. Suggest simplifying this paragraph
```
looks good, can delete the temporary comment

jwwojak replied:
Will remove the comment.

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -79,12 +85,39 @@ You can include any other information you like in the metadata dictionary. The f
+^^^^^^^^^^^^
+.. My understanding of the requirements code block. Needs verification.
+
+The ``requirements`` code block appears after the ``metadata`` code block in a Python protocol. It defines the robot model (e.g. ``Flex`` or ``OT-2``) and the API version you're using.
```
technically these blocks can go just about anywhere. but this is good form, so saying "it goes here" is 👍

jwwojak replied:
Max suggested putting the `requirements` first, before `metadata` but also indicated order doesn't matter.

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -79,12 +85,39 @@ You can include any other information you like in the metadata dictionary. The f
+    # Flex
+    requirements = {"robotType": "Flex", "apiLevel": "2.15"}
+
+OT-2 users working with API v2.15 (or higher) also need to specify the API version in the ``requirements`` instead of in the ``metadata``. For example:
```
as of right now, you can specify any `apiLevel` in metadata. the only required key in requirements is `robotType`. but we should check if that's the long-term plan.

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -95,43 +128,75 @@ With the protocol context argument named and typed, you can start calling method
+           :align: center
+           :alt: OT-2 deck map with a tip rack in slot 1, reservoir in slot 2, and well plate in slot 3.
+
+You may notice that these deck maps don't show where the liquids will be at the start of the protocol. Liquid definitions aren’t required in Python protocols, unlike protocols made in `Protocol Designer <https://designer.opentrons.com/>`_. (Sneak peek: you’ll put the diluent in column 1 of the reservoir and the solution in column 2 of the reservoir.)
```
i think it's too advanced to put liquid defs in the tutorial, but we should link to that section of the Python docs. (it didn't exist when the Tutorial was first published!)

jwwojak replied:
Good point! Will revise and update with [Labelling Liquids in Wells](https://docs.opentrons.com/v2/new_labware.html#labeling-liquids-in-wells).

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -95,43 +128,75 @@ With the protocol context argument named and typed, you can start calling method
 --------
 
-Next you’ll specify what pipette to use in the protocol. Loading a pipette is done with the :py:meth:`.load_instrument` method, which takes three arguments: the name of the pipette, the mount it’s installed in, and the tip racks it should use when performing transfers. Load whatever pipette you have installed in your robot by using its :ref:`standard pipette name <new-pipette-models>`. Here’s how ...
+Next you’ll specify what pipette to use in the protocol. Loading a pipette is done with the :py:meth:`.load_instrument` method, which takes three arguments: the name of the pipette, the mount it’s installed in, and the tip racks it should use when performing transfers. Load whatever pipette you have installed in your robot by using its :ref:`standard pipette name <new-pipette-models>`. Here’s how ...
```
```suggestion
Next you’ll specify what pipette to use in the protocol. Loading a pipette is done with the :py:meth:`.load_instrument` method, which takes three arguments: the name of the pipette, the mount it’s installed in, and the tip racks it should use when performing transfers. Load whatever pipette you have installed in your robot by using its :ref:`standard pipette name <new-pipette-models>`. Here’s how to load the pipette in the left mount:
```
can simplify

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -95,43 +128,75 @@ With the protocol context argument named and typed, you can start calling method
         p300 = protocol.load_instrument('p300_single_gen2', 'left', tip_racks=[tips])
 
-Since the pipette is so fundamental to the protocol, it might seem like you should have specified it first. But there’s a good reason why pipettes are loaded after labware: you need to have already loaded ``tips`` in order to tell the pipette to use it. And now you won’t have to reference ``tips`` again in your code — it’s assigned to the ``p300`` pipette and the robot will know to use it when co ...
+Since the pipette is so fundamental to the protocol, it might seem like you should have specified it first. But there’s a good reason why pipettes are loaded after labware: you need to have already loaded ``tips`` in order to tell the pipette to use it. And now you won’t have to reference ``tips`` again in your code — it’s assigned to the ``p1000`` (Flex) or ``p300`` (OT-2) pipette and the robot  ...
```
if you think explaining the different variable names like this is awkward, we could use something like `left_pipette` instead for both robots.

ecormany replied:
this will have the added benefit of letting some of the code snippets below be robot-agnostic, instead of having to provide two one-line snippets that don't really differ in function at all.

jwwojak replied:
Ah, now I understand. Yes, instead of `p300` or `p1000`, let's use `left_pipette` for the variable name. And this transfers over to the animated gif. Will update.

jwwojak replied:
Updated with `left_pipette`

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -147,9 +212,17 @@ Let’s start with the diluent. This phase takes a larger quantity of liquid and
-Breaking down this single line of code shows the power of :ref:`complex commands <v2-complex-commands>`. The first argument is the amount to transfer to each destination, 100 µL. The second argument is the source, column 1 of the reservoir (which is still specified with grid-style coordinates as ``A1`` — a reservoir only has an A row). The third argument is the destination. Here, calling the :py: ...
+Breaking down these single lines of code shows the power of :ref:`complex commands <v2-complex-commands>`. The first argument is the amount to transfer to each destination, 100 µL. The second argument is the source, column 1 of the reservoir (which is still specified with grid-style coordinates as ``A1`` — a reservoir only has an A row). The third argument is the destination. Here, calling the :p ...
+
+.. Possible short text to explain the animation uses OT-2 code but applies to Flex as well?
```
ah, dang. i have the source file and could fairly easily update the animations.

jwwojak replied:
It isn't too late. Do you want to update the animation? 

Maybe that isn't required. Even though the code here uses the OT-2 pipette variable name, the concept is the same for both robots.  Creating 2 separate image files may be more work than needed to convey the idea showing stepwise transfer of liquids. Instead, we could try some disclaimer text like this:

"The following animated images use OT-2 code, but the concepts apply to how a Flex and OT-2 transfer liquids in this tutorial."

Something like that maybe? 

There are 5 animated gifs here. Making separate images for Flex creates a total of 10 images and also multiple tabs (if we put images in those). That's a lot of extra files to maintain and more tabs may make the page look busy.

ecormany replied:
i meant that if we went with a single pipette variable name, e.g. `left_pipette`, i could make one set of updated images. same overall complexity. just updating that little bit of text and re-exporting.

jwwojak replied:
Fixing this now. Change `p300` and `p1000` to `left_pipette` so we can use one code example for both robots and the animated image.

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -244,7 +343,7 @@ If you get any errors in simulation, or you don't get the outcome you expected w
 ^^^^^^^^^^^^^
 
-Simulation doesn’t require having an OT-2 connected to your computer. You just need to install the ``opentrons`` Python module using ``pip``. This will give you access to the ``opentrons_simulate`` command-line utility (``opentrons_simulate.exe`` on Windows). To see a text preview of the steps the OT-2 will take, ``cd`` to the directory where you saved your protocol file and run:
+Simulation doesn’t require having a robot to your computer. You just need to install the ``opentrons`` Python module using ``pip``. This will give you access to the ``opentrons_simulate`` command-line utility (``opentrons_simulate.exe`` on Windows). To see a text preview of the steps the OT-2 will take, ``cd`` to the directory where you saved your protocol file and run:
```
```suggestion
Simulation doesn’t require having a robot connected to your computer. You just need to install the ``opentrons`` Python module using ``pip``. This will give you access to the ``opentrons_simulate`` command-line utility (``opentrons_simulate.exe`` on Windows). To see a text preview of the steps the OT-2 will take, ``cd`` to the directory where you saved your protocol file and run:
```

ecormany commented on api/docs/v2/tutorial.rst:
```diff
@@ -263,7 +362,7 @@ If that’s too long, you can always cancel your run partway through or modify `
 ^^^^^^^^^^
 
-The simplest way to run your protocol on an OT-2 is to use the `Opentrons App <https://opentrons.com/ot-app>`_. Once you’ve installed the app and connected to your robot, navigate to the **Protocol** tab. Click **Choose File…** and open your protocol from the file picker. You should see “Protocol - Serial Dilution Tutorial” (or whatever ``protocolName`` you entered in the metadata) in a banner at ...
+The simplest way to run your protocol on a Flex or OT-2 is to use the `Opentrons App <https://opentrons.com/ot-app>`_. Once you’ve installed the app and connected to your robot, navigate to the **Protocol** tab. Click **Choose File…** and open your protocol from the file picker. You should see “Protocol - Serial Dilution Tutorial” (or whatever ``protocolName`` you entered in the metadata) in a ba ...
```
on line 369 below: `Protocol Steps` are now called `Run Preview`. we should change the term here.

jwwojak replied:
Ok. Will change to `Run Preview`.

ecormany reviewed (changes requested):
This is good progress towards getting the Tutorial working for both Flex and OT-2! I think in general it's best to make it so when we have different code snippets for different robots, it's due to a meaningful difference — not just a different pipette variable name. Once we have all the snippets sorted, we should update the existing full protocol files and add Flex variants to the Try Your Protocol section.

jwwojak committed:
Update api/docs/v2/tutorial.rst

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Update api/docs/v2/tutorial.rst

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Update tutorial.rst

Removing commented text

jwwojak committed:
Update tutorial.rst

Removing "opentrons import" in favor of "import"

jwwojak committed:
Update api/docs/v2/tutorial.rst

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Update tutorial.rst

jwwojak committed:
Update tutorial.rst

jwwojak committed:
Update tutorial.rst

Changes based on reviewer feedback.

Also, trying the labware in a table, rather than using nested bullets.

We ought to update the serial dilution code sample files on github.

jwwojak committed:
Merge branch 'edge' into tutorial-page-revisions

jwwojak committed:
Update tutorial.rst

More changes:

- Labware bullet list changed to table
- Move water out of labware list into own sentence
- Added some optional links to resources mentioned in text but not explained (e.g. pip, CD command)
- Other small changes that may be helpful
- Will need to update code samples mentioned in "Try Your Protocol" section on git
- Still uncertainty about the `requirements` code block

Ready for more review.

ecormany committed:
update images, fix indenting

ecormany committed:
update links and terminal commands

jwwojak committed:
Update tutorial.rst

Updated the "Requirements" section. Make it sound more optional, but a really good idea instead of required.

ecormany committed:
proofread, 5.x->6.x+ interface descriptions

ecormany reviewed (approved):
Made a couple final proofreading changes and updated some descriptions of the app UX to reflect the current version. :shipit:

ecormany committed:
oops one last apiLevel fix
